### PR TITLE
winmidi: SysEx fixes

### PR DESCRIPTION
**Pull request by ceski-1**
This is a draft for now. Let's make this thing even better @rfomin .

### CODE REVIEW

**Comment by ceski-1** [comment]
1. Update steps for resetting device
2. Fix very old typos for `midi_controller_t` values in midifile.h (CC#3 is undefined)
3. Add additional CC#'s to `midi_controller_t`

**Comment by ceski-1** [comment]
Configurable delay after reset works by tracking tempo and timediv in order to convert a user-defined delay in ms to a delay in ticks. A NOP messasge is sent with the tick delay immediately after a SysEx reset. Even if the user doesn't use a SysEx reset, some devices may need a delay to finish processing the channel messages. This is how MIDI songs are written as well.

Default delay after reset is 100 ms:
- MIDI-OX uses 60 ms for SysEx messsages
- FSMP uses 20 + 40 = 60 ms for SysEx messages
- Roland requires ~50 ms for a "GM System On"
- Yamaha requires ~50 ms for a "GM System On"
- MIDI specs say that industry median time for a "GM System On" was 100 ms in the year it was published (1996)

References:

Roland SC-55mkii manual page 93:
> Turn General MIDI System On
It takes about 50ms to execute this message.

Yamaha MU80 manual (E2) page 11:
> 2.1.3.1 General MIDI Mode On
This message requires approximately 50ms to execute, so sufficient time should be allowed
before the next message is sent.

Complete MIDI 1.0 Detailed Specification v. 96.1 3rd Ed. page 4:
> Description of Issue:
After receiving a “GM System On” message, some devices need a period of time to reset themselves before they can start producing sound. How prevalent is this practice, what length pauses are required, and what can composers and authors do to avoid problems? In addition, some devices will go in or out of GM mode under certain conditions. What are these conditions, and are any of them  acceptable?
Findings:
5 of the hardware respondents said that their devices don’t need any amount of time after receiving a Turn GM System On message before they can play sound. 2 were vague about the length of the pause needed, and of the others the range was from 10 ms all the way to “1-2 seconds”, with 100 ms being the median value.

**Comment by ceski-1** [comment]
The last update fixes VirtualMIDISynth for good.

**Comment by ceski-1** [comment]
SysEx messages from MIDI files are blocked by default:
- Vanilla Doom doesn't process SysEx messages
- Source ports don't process SysEx messages (except for my PrBoom+ changes)
- MIDI files specifically for Doom wads were written with the above in mind
- MIDI files were ripped from other games/composers without considering the SysEx messages they contained (see https://github.com/fabiangreffrath/crispy-doom/issues/940#issuecomment-1238260970)

Robust SysEx message detection will be added later to make it safe to allow SysEx messages from MIDI files.

File: src/i_winmusic.c [hunk 1]
```diff
@@ -376,8 +430,16 @@ static void FillBuffer(void)
...
-                    MEVT_LONGMSG);
+                if (winmm_allow_sysex)
+                {
+                    uint32_t length = event->data.sysex.length + sizeof(byte);
+                    data = MAKE_EVT(length, 0, 0, MEVT_LONGMSG);
+                }
+                else
+                {
+                    // Preserve timing with a NOP.
+                    data = MAKE_EVT(0, 0, 0, MEVT_NOP);
```
Comment (rfomin) [inline src/i_winmusic.c:441]: This line causes the JJ2HELL track from MusChange.wad to not play.

**Comment by rfomin** [comment]
Overall this looks great, thank you!

> Robust SysEx message detection will be added later to make it safe to allow SysEx messages from MIDI files.

It's necessary? I have tested several XG MIDI and they work great. Maybe we can enable SysEx messages by default, we had a request from a modder/composer: [[doomworld]](https://www.doomworld.com/forum/topic/66518-doom-retro-v462-updated-september-1-2022/?do=findComment&comment=2571424) Also what do you think about loops support? [[doomworld]](https://www.doomworld.com/forum/topic/66518-doom-retro-v462-updated-september-1-2022/?do=findComment&comment=2574277)

**Comment by ceski-1** [comment]
> It's necessary? I have tested several XG MIDI and they work great.
Yes. Let me give an example. If we blindly let every SysEx message through without some basic checking, the MIDI file could send a SysEx reset and set every channel volume back to 100 and hurt your ears. So at the bare minimum, I will be adding SysEx reset checks just like PrBoom+. I wish `midiOutSetVolume()` worked still. We could avoid all these headaches.
> Maybe we can enable SysEx messages by default, we had a request from a modder/composer: [[doomworld]](https://www.doomworld.com/forum/topic/66518-doom-retro-v462-updated-september-1-2022/?do=findComment&comment=2571424)
After I get the checks in place, SysEx can be re-enabled.
> Also what do you think about loops support? [[doomworld]](https://www.doomworld.com/forum/topic/66518-doom-retro-v462-updated-september-1-2022/?do=findComment&comment=2574277)
I think it would worth checking out. I believe EDuke32 supports custom loop points. It would be something to look at after I address everything else first. Maybe have them open an issue for the feature, or do it for them?

**Comment by ceski-1** [comment]
> I propose this change to send all messages in the player thread. Apparently, `FillBuffer()` can sometimes cause a slight visual lag (`midiStreamOut()` lags when it sends a lot of messages?)

How about using: `SetEvent(hBufferReturnEvent);`

**Comment by rfomin** [comment]
> How about using: `SetEvent(hBufferReturnEvent);`

Yes, this is much better.

> I believe EDuke32 supports custom loop points.

EDuke32 and ZMusic support EMIDI extensions: https://moddingwiki.shikadi.net/wiki/Apogee_Expanded_MIDI

**Comment by ceski-1** [comment]
I'm going to try to keep these settings "live" or "active"? Not sure the term. It's so we can potentially have them in the menus as options that can change without having to restart Woof.

**Comment by rfomin** [comment]
> I'm going to try to keep these settings "live" or "active"? Not sure the term. It's so we can potentially have them in the menus as options that can change without having to restart Woof.

I think then we should at least restart the song, so it's not necessary.

**Review by rfomin — APPROVED** [review]
Looks good to me, thank you!

File: src/m_misc.c [hunk 2]
```diff
@@ -2319,6 +2322,20 @@ default_t defaults[] = {
 #endif
 
 #if defined(_WIN32)
+  {
```
Comment (fabiangreffrath) [inline src/m_misc.c:2325]: Disclaimer: I have no idea what you guys are doing here.

But, do we really need another three config variables? I mean, I'd consider myself a rather "advanced user", but still I have absolutely no clue what these are all about. How should the avarage user know what's the right choice here and what consequence a bad choice would have? Is there anything like a reasonable default that applies everywhere?

**Review by fabiangreffrath — APPROVED** [review]
Code wise it's all fly, but I feel like the sheer number of configuration options this adds is growing over our heads. But maybe that's just me...

File: src/m_misc.c (hunk 2 above)
Comment (rfomin) [reply src/m_misc.c:2325]: This is for MIDI enthusiasts - you need to buy a vintage sound card (Roland SC-55 or something) and make it work on a modern system, I don't know how.
What I've done:
1. Install the Windows MIDI Mapper selection tool (VirtualMIDISynth)
2. Install the loopMIDI driver.
3. Set up a simple VST host or use a DAW.
4. Get some VST MIDI synthesizer plugin. Roland has a [Sound Canvas VA](https://www.roland.com/us/products/rc_sound_canvas_va/) product, I recommend the [Yamaha S-YXG50](https://veg.by/en/projects/syxg50/) ("abandonware").

Once you install and configure all of this, these settings will start to make sense 😄 Maybe we should even add a SysEx option to the menu. We develop software for enthusiasts and we have them in our DW thread.

Comment (fabiangreffrath) [reply src/m_misc.c:2325]: Okay, this is clearly your (plural) ground and I rely on your expertise. I'm out.
giphy

**Pull request merged**